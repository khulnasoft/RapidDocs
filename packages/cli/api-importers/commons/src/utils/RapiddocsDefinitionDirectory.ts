import { RawSchemas } from "@khulnasoft/rapiddocs-definition-schema";
import { RelativeFilePath, sep } from "@khulnasoft/path-utils";

export class RapiddocsDefinitionDirectory {
    private files: Record<RelativeFilePath, RawSchemas.DefinitionFileSchema> = {};
    private directories: Record<string, RapiddocsDefinitionDirectory> = {};

    public getAllFiles(): Record<RelativeFilePath, RawSchemas.DefinitionFileSchema> {
        const files: Record<RelativeFilePath, RawSchemas.DefinitionFileSchema> = {};

        const walk = (root: RapiddocsDefinitionDirectory, currentPath?: string) => {
            for (const [relativeFilePath, definition] of Object.entries(root.files)) {
                const fullRelativeFilePath =
                    currentPath != null
                        ? RelativeFilePath.of(`${currentPath}${sep}${relativeFilePath}`)
                        : RelativeFilePath.of(relativeFilePath);
                files[fullRelativeFilePath] = definition;
            }
            const sortedDirectories = Object.keys(root.directories).sort();
            for (const directory of sortedDirectories) {
                const nextPath = currentPath != null ? `${currentPath}${sep}${directory}` : directory;
                // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
                const nextDirectory = root.directories[directory]!;
                walk(nextDirectory, nextPath);
            }
        };

        walk(this);

        return files;
    }

    public getOrCreateFile(relativeFilePath: RelativeFilePath): RawSchemas.DefinitionFileSchema {
        return this.getOrCreateFileRecursive(relativeFilePath.split(sep));
    }

    private getOrCreateFileRecursive(pathParts: string[]): RawSchemas.DefinitionFileSchema {
        if (pathParts.length === 1) {
            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
            return (this.files[RelativeFilePath.of(pathParts[0]!)] ??= {});
        }
        const [directory, ...remainingPath] = pathParts;
        if (directory == null) {
            throw new Error(`Internal error; cannot add file with path: ${pathParts}`);
        }
        if (!this.directories[directory]) {
            this.directories[directory] = new RapiddocsDefinitionDirectory();
        }
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        return this.directories[directory]!.getOrCreateFileRecursive(remainingPath);
    }
}
