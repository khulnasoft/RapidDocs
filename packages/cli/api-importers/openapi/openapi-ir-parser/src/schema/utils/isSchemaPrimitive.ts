import { Schema } from "@khulnasoft/openapi-ir";

export function isSchemaPrimitive(schema: Schema): boolean {
    if (schema.type === "primitive") {
        return true;
    } else if (schema.type === "nullable" || schema.type === "optional") {
        return isSchemaPrimitive(schema.value);
    }
    return false;
}
