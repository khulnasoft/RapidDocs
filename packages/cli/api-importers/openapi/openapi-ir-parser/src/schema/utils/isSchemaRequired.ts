import { Schema } from "@khulnasoft/openapi-ir";

export function isSchemaRequired(schema: Schema): boolean {
    return schema.type !== "optional" && schema.type !== "nullable";
}
