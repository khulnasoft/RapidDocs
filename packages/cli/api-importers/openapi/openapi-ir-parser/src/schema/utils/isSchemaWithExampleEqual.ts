import { SchemaWithExample, isSchemaEqual } from "@khulnasoft/openapi-ir";

import { convertSchemaWithExampleToSchema } from "./convertSchemaWithExampleToSchema";

// only diffs the schema, not the example
export function isSchemaWithExampleEqual(a: SchemaWithExample, b: SchemaWithExample): boolean {
    return isSchemaEqual(convertSchemaWithExampleToSchema(a), convertSchemaWithExampleToSchema(b));
}
