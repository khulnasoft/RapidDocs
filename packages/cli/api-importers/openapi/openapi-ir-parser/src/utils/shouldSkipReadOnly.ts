import { RapiddocsOpenapiIr } from "@khulnasoft/openapi-ir";

export function shouldSkipReadOnly(method: RapiddocsOpenapiIr.HttpMethod): boolean | undefined {
    return method === "POST" || method === "PUT" || method === "PATCH";
}
