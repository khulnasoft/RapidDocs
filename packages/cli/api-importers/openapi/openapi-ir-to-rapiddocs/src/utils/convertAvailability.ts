import { RawSchemas } from "@khulnasoft/rapiddocs-definition-schema";
import { Availability } from "@khulnasoft/openapi-ir";

export function convertAvailability(
    availability: Availability | undefined
): RawSchemas.AvailabilityUnionSchema | undefined {
    switch (availability) {
        case Availability.Deprecated:
            return "deprecated";
        case Availability.Beta:
            return "pre-release";
        case Availability.GenerallyAvailable:
            return "generally-available";
        default:
            return undefined;
    }
}
