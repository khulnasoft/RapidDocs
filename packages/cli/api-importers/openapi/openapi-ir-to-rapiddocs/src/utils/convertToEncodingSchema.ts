import { assertNever } from "@khulnasoft/core-utils";
import { RawSchemas } from "@khulnasoft/rapiddocs-definition-schema";
import { Encoding } from "@khulnasoft/openapi-ir";

export function convertToEncodingSchema(encoding: Encoding): RawSchemas.EncodingSchema {
    switch (encoding.type) {
        case "protobuf":
            return {
                proto: {
                    type: encoding.typeName
                }
            };
        default:
            assertNever(encoding.type);
    }
}
