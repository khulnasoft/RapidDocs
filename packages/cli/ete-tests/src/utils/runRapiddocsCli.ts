import { Options } from "execa";
import path from "path";

import { loggingExeca } from "@khulnasoft/logging-execa";

export async function runRapiddocsCli(args: string[], options?: Options): Promise<loggingExeca.ReturnValue> {
    return loggingExeca(undefined, "node", [path.join(__dirname, "../../../cli/dist/dev/cli.cjs"), ...args], {
        ...options,
        env: {
            ...options?.env,
            RAPIDDOCS_TOKEN: process.env.RAPIDDOCS_ORG_TOKEN_DEV
        },
        doNotPipeOutput: options?.reject === false
    });
}
