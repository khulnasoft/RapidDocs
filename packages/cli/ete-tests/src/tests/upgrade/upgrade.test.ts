import { readFile, writeFile } from "fs/promises";
import yaml from "js-yaml";

import { RAPIDDOCS_DIRECTORY, GENERATORS_CONFIGURATION_FILENAME, generatorsYml } from "@khulnasoft/configuration";
import { RelativeFilePath, join } from "@khulnasoft/fs-utils";

import { runRapiddocsCli } from "../../utils/runRapiddocsCli";
import { init } from "../init/init";

const GENERATORS_CONFIGURATION: generatorsYml.GeneratorsConfigurationSchema = {
    groups: {
        internal: {
            generators: [
                {
                    name: "rapiddocsapi/rapiddocs-postman",
                    version: "0.0.15",
                    output: {
                        location: "local-file-system",
                        path: "./generated-postman"
                    }
                },
                {
                    name: "rapiddocsapi/rapiddocs-openapi",
                    version: "0.0.3",
                    output: {
                        location: "local-file-system",
                        path: "./generated-openapi"
                    },
                    config: {
                        format: "yaml"
                    }
                },
                {
                    name: "rapiddocsapi/rapiddocs-java-sdk",
                    version: "0.0.50",
                    output: {
                        location: "maven",
                        coordinate: ""
                    }
                },
                {
                    name: "rapiddocsapi/rapiddocs-typescript-sdk",
                    version: "0.0.11",
                    output: {
                        location: "npm",
                        "package-name": ""
                    }
                }
            ]
        },
        external: {
            generators: [
                {
                    name: "rapiddocsapi/rapiddocs-postman",
                    version: "0.0.20",
                    output: {
                        location: "local-file-system",
                        path: "./generated-postman"
                    }
                },
                {
                    name: "rapiddocsapi/rapiddocs-openapi",
                    version: "0.0.2",
                    output: {
                        location: "local-file-system",
                        path: "./generated-openapi"
                    },
                    config: {
                        format: "yaml"
                    }
                },
                {
                    name: "rapiddocsapi/rapiddocs-java-sdk",
                    version: "0.0.81",
                    output: {
                        location: "maven",
                        coordinate: ""
                    }
                },
                {
                    name: "rapiddocsapi/rapiddocs-typescript-sdk",
                    version: "0.0.14",
                    output: {
                        location: "npm",
                        "package-name": ""
                    }
                }
            ]
        }
    }
};

describe("rapiddocs upgrade", () => {
    it("upgrades generators", async () => {
        const directory = await init();
        const generatorsConfigurationFilepath = join(
            directory,
            RelativeFilePath.of(RAPIDDOCS_DIRECTORY),
            RelativeFilePath.of(GENERATORS_CONFIGURATION_FILENAME)
        );
        // make sure the file exists
        await readFile(generatorsConfigurationFilepath);
        await writeFile(generatorsConfigurationFilepath, yaml.dump(GENERATORS_CONFIGURATION));
        await runRapiddocsCli(["upgrade"], {
            cwd: directory,
            env: {
                // this env var needs to be defined so the CLI thinks we're mid-upgrade
                RAPIDDOCS_PRE_UPGRADE_VERSION: "0.0.0"
            }
        });
        const generatorsConfiguration = (await readFile(generatorsConfigurationFilepath)).toString();
        expect(generatorsConfiguration).toMatchSnapshot();
    }, 90_000);
});
