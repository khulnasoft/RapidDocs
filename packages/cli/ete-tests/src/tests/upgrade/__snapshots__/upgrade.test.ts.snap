// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`rapiddocs upgrade > upgrades generators 1`] = `
"groups:
  internal:
    generators:
      - name: rapiddocsapi/rapiddocs-postman
        version: 0.0.15
        output:
          location: local-file-system
          path: ./generated-postman
      - name: rapiddocsapi/rapiddocs-openapi
        version: 0.0.3
        output:
          location: local-file-system
          path: ./generated-openapi
        config:
          format: yaml
      - name: rapiddocsapi/rapiddocs-java-sdk
        version: 0.0.50
        output:
          location: maven
          coordinate: ''
      - name: rapiddocsapi/rapiddocs-typescript-sdk
        version: 0.0.11
        output:
          location: npm
          package-name: ''
  external:
    generators:
      - name: rapiddocsapi/rapiddocs-postman
        version: 0.0.20
        output:
          location: local-file-system
          path: ./generated-postman
      - name: rapiddocsapi/rapiddocs-openapi
        version: 0.0.2
        output:
          location: local-file-system
          path: ./generated-openapi
        config:
          format: yaml
      - name: rapiddocsapi/rapiddocs-java-sdk
        version: 0.0.81
        output:
          location: maven
          coordinate: ''
      - name: rapiddocsapi/rapiddocs-typescript-sdk
        version: 0.0.14
        output:
          location: npm
          package-name: ''
"
`;
